### ISSUE TITLE
Leveldb memory usage during compaction

### ISSUE THREAD

**Issue by gmather** [lovoo/goka#209]
Hello,
I'm seeing a spike in heap after a large Persist topic on a processor is recovered without an existing leveldb. From what I can tell, this recovery is happening via a leveldb transaction to pause compaction and once recovered it'll commit and cause an auto-compaction. From profiling the heap, there is GBs allocated for this. Once fully recovered the processor runs at steady state in the 100s of MBs.

Do you have some insight into how the leveldb options can help control the memory footprint or is this expected when starting a fresh leveldb with an existing topic?

**Comment by db7** [comment]
You analysis is correct. Recovery happens in a transaction and once the transaction is done, the compaction kicks in, and that may take a while. AFAIR the recovery turns out to be faster in this way than doing compaction all the time. The drawback is the memory spike.

You can also remove the transaction in the implementation of the storage. You can actually build your own storage and pass it to the processors/views as argument.

Regarding options for LevelDB, there were some option examples here: https://github.com/lovoo/goka/wiki/Tips#configuring-leveldb

Perhaps somebody from LOVOO (@frairon @TheL1ne) can share their actual configuration for some large topic?

**Comment by frairon** [comment]
Hi @gmather, as Diogo pointed out the behavior you experience is the expected but I admit it's not optimal regarding memory consumption.
We did try out regular transaction commits on leveldb but as said it was a bit slower so we opted for spending the memory. 
Our processors are deployed in Kubernetes and our (dirty) solution for this problem is simply to allow the pod to use much more memory than during normal operation.
We don't have a magic leveldb configuration that would help here, sorry.

One workaround/solution I could think of could a sequential transaction commit. So whenever a partition is done recovering it would wait until the previous partition is done commitng. But that would require to add an option and some synchronization mechanism. If it's a critical issue for you we can look into that option.
